src/resolvers/git.cr

```diff
@@ -100,12 +100,25 @@ module Shards
       "git"
     end
 
+    private KEY_HOSTS = {"www.github.com", "github.com", "www.bitbucket.com", "bitbucket.com", "www.gitlab.com", "gitlab.com"}
+
     def self.normalize_key_source(key : String, source : String) : {String, String}
       case key
       when "git"
-        {"git", source}
+        uri = URI.parse(source)
+        downcased_host = uri.host.try &.downcase
+        scheme = uri.scheme.try &.downcase
+        if scheme.in?("git", "http", "https") && downcased_host && downcased_host.in?(KEY_HOSTS)
+          uri.scheme = "https"
+          downcased_path = uri.path.downcase
+          uri.path = downcased_path.ends_with?(".git") ? downcased_path : "#{downcased_path}.git"
+          uri.host = downcased_host.gsub("www.", "")
```

Comment:
ditto, again I don't think we should transform the path in this way, as it will hide the errors on the user side. This PR is concerned about resolving repositories in a case insensitive manner, not "fixing" the repo paths for users.

Reply:
github, gitlab and bitbucket all default their clone URLs without www and redirect to to this version of the URL and enforce HTTPS transfer.

So by doing this not only are we preventing potential clashes, we're also improving performance.

<img width="354" alt="image" src="https://user-images.githubusercontent.com/368013/106403996-3f475700-6485-11eb-9c2f-762535118ff8.png">

<img width="343" alt="image" src="https://user-images.githubusercontent.com/368013/106404017-54bc8100-6485-11eb-95b3-e9dc862bc501.png">

Reply:
That's correct, but I don't get why want you handle this within generic `git` resolver?